### ISSUE TITLE
Exception: Fetch error, code == 403

### ISSUE THREAD

**Issue by rohshr** [kallqvist/skillshare-downloader#32]
I have a valid subscription but I have been receiving this error. This is happening after I had successfully downloaded a couple of courses.

**Comment by ravinot** [comment]
@rohshr I believe this is happening across a number of scripts that do similar things. It looks like something may have changed with the way authentication tokens are handled, maybe on the Skillshare end.

**Comment by houdinii** [comment]
I don't know a whole lot of JS or really where to look to start unraveling this. How was the endpoint found the first time?

**Comment by ravinot** [comment]
The issue may be due to the macOS version being run. I have Catalina on one computer and High Sierra on another. High Sierra works fine.

**Comment by kallqvist** [comment]
Sorry for delay guys, I've finally had some time to circle back to this during weekend.

@houdinii Last time around I just did a whole bunch of fiddling with the dev console in chrome. Browser is doing all sign-in stuff and I just copied the cookie value out again after it's done

@ravinot Is that the exact same cookie value you use in the code on both your computers or did you do the browser sign-in one time per computer? And it's the same skillshare account you use on both computers, right?

I've tried running the script here with my (non-premium) skillshare account and it did work as it usually does (that is that the first "free" lesson-overview gets downloaded OK and then it fails for all following parts).
I'm not really able to see anything different with either cookies or APIs here so am a bit confused...

Not sure how well skillshare API is following the HTTP standard but normally http 403 means that cookie is understood and that the authentication is OK but the service for some other reason is refusing to fullfil the request.

**Comment by ravinot** [comment]
@kallqvist Appreciate the reply. I did the browser sign-in one time per computer, using the same Skillshare account. I was stumped why the script was working and then stopped on my machine, and the only things that I could think of that changed recently on my Mac was the OS and the version of Python I had installed. My comment was just my personal experience of what could be a resolution to the issue @rohshr initially mentioned, but I realise that each person is going to have their own unique development environment.

Thanks for even writing the script in the first place!

**Comment by rohshr** [comment]
So after tracing the request through Postman, it turned out that the request was being denied because of Cloudflare. Using the [cloudscraper](https://github.com/venomous/cloudscraper) module to replace the regular requests seemed to fix the issue for me. Not sure why the requests were getting blocked in the first place though.

**Comment by kallqvist** [comment]
Hmm, interesting information @rohshr !
I'm quite busy with other work right now but hope I'll get to reviewing and merging it into main soon-ish...
I haven't been able to reproduce the issue myself but will leave this issue open for now and drop the link to your pull-request here if anybody else is in need of a workaround: https://github.com/kallqvist/skillshare-downloader/pull/33

**Comment by kallqvist** [comment]
PS. if this is the issue everyone else have been experiencing I'm guessing it has nothing to do with skillshare accounts or cookies etc at all as I thought at first. More likely I think that it's some kind of throttling or "anti-spam / anti-ddos" or similar that triggers in that case. Possibly it happens when many videos have been downloaded from the same IP-adress within some period of time or such...

**Comment by Dellgenius** [comment]
@rohshr I tried your fix but I still seem to be getting the 403 error for some reason.

**Comment by windowswithoutborders** [comment]
Experiencing the 403 error with the main branch as well as rohshr's fork. For what it's worth I'm executing the script on python 3.8.2.

**Comment by Risath** [comment]
Hi All, I managed to fix this issue:

Just replace the header values at https://github.com/kallqvist/skillshare-downloader/blob/master/code/downloader.py#L108
with this:			
	
```
            headers={
                'Referer': 'https://www.skillshare.com/',
                'cookie': self.cookie,
            }
```

**Comment by silverbret1709** [comment]
> Hi All, I managed to fix this issue:
> 
> Just replace the header values at https://github.com/kallqvist/skillshare-downloader/blob/master/code/downloader.py#L108
> with this:
> 
> ```
>             headers={
>                 'Referer': 'https://www.skillshare.com/',
>                 'cookie': self.cookie
>             }
> ```

It works, Thank you very much

**Comment by silverbret1709** [comment]
> Hi All, I managed to fix this issue:
> 
> Just replace the header values at https://github.com/kallqvist/skillshare-downloader/blob/master/code/downloader.py#L108
> with this:
> 
> ```
>             headers={
>                 'Referer': 'https://www.skillshare.com/',
>                 'cookie': self.cookie
>             }
> ```
Some Course i still get error:  Failed to read video ID from data

**Comment by houdinii** [comment]
I kept the User-Agent as 'User-Agent': 'Android 5.1.1' without the SkillShare app info and added the above Referer info, and that also allowed things to work as well, so I think it's just the bit about the app making things wonky. I have not had an issue with failing to read video ID yet, but I just started. 

```
headers={
                'User-Agent': 'Android 5.1.1',
                'Referer': 'https://www.skillshare.com/',
                'cookie': self.cookie,
            }
```

**Comment by faqu** [comment]
> I kept the User-Agent as 'User-Agent': 'Android 5.1.1' without the SkillShare app info and added the above Referer info, and that also allowed things to work as well, so I think it's just the bit about the app making things wonky. I have not had an issue with failing to read video ID yet, but I just started.
> 
> ```
> headers={
>                 'User-Agent': 'Android 5.1.1',
>                 'Referer': 'https://www.skillshare.com/',
>                 'cookie': self.cookie,
>             }
> ```

It seems that the Skillshare's server prevents some unwanted headers to get access to the server.

**Comment by silverbret1709** [comment]
> I kept the User-Agent as 'User-Agent': 'Android 5.1.1' without the SkillShare app info and added the above Referer info, and that also allowed things to work as well, so I think it's just the bit about the app making things wonky. I have not had an issue with failing to read video ID yet, but I just started.
> 
> ```
> headers={
>                 'User-Agent': 'Android 5.1.1',
>                 'Referer': 'https://www.skillshare.com/',
>                 'cookie': self.cookie,
>             }
> ```

Still error with me :(

**Comment by faqu** [comment]
Try to change it like this:
>            headers={
>                'Accept': 'application/vnd.skillshare.class+json;,version=0.8',
>                'User-Agent': 'Skillshare/5.3.0; Android 9.0.1',
>                'Host': 'api.skillshare.com',
>                'Referer': 'https://www.skillshare.com/',
>                'cookie': self.cookie,
>            }

I just tested and it worked.

**Comment by silverbret1709** [comment]
It worked, thank you (❁´◡`❁)

**Comment by AmirGhiasi95** [comment]
> Try to change it like this:
> 
> > ```
> >        headers={
> >            'Accept': 'application/vnd.skillshare.class+json;,version=0.8',
> >            'User-Agent': 'Skillshare/5.3.0; Android 9.0.1',
> >            'Host': 'api.skillshare.com',
> >            'Referer': 'https://www.skillshare.com/',
> >            'cookie': self.cookie,
> >        }
> > ```
After implementing these lines, it started working but in an attempt to resume an unfinished download, which normally should skip already downloaded videos, it skips 20 videos and results in multiple errors.

`Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond'`

**Comment by mahasatwa** [comment]
> After implementing these lines, it started working but in an attempt to resume an unfinished download, which normally should skip already downloaded videos, it skips 20 videos and results in multiple errors.
> 
> `Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond'`

This should be on a different topic/issue.

The lines above have nothing to do with the error you occur. 
Mind to take a little search on StackOverflow or anywhere about these errors.

I just search it on google, and on the first page, this [StackOverflow answer](https://stackoverflow.com/questions/27752644/python-requests-error-10060/27753345#27753345) might answer your issue.
You need to set a `timeout` parameter for `requests.get` on [line 152](https://github.com/kallqvist/skillshare-downloader/blob/master/code/downloader.py#L108) like this:
>            response = requests.get(dl_url, allow_redirects=True, stream=True, timeout=15)
if the error still occur, might be your environment network problem.

**Comment by mattl202** [comment]
> Try to change it like this:
> 
> > ```
> >        headers={
> >            'Accept': 'application/vnd.skillshare.class+json;,version=0.8',
> >            'User-Agent': 'Skillshare/5.3.0; Android 9.0.1',
> >            'Host': 'api.skillshare.com',
> >            'Referer': 'https://www.skillshare.com/',
> >            'cookie': self.cookie,
> >        }
> > ```
> 
> I just tested and it worked.

Worked for me, thank you

**Comment by rupaniraheel** [comment]
can anybody tell me how to set up that scripts for download the videos from skillshare

**Comment by kallqvist** [comment]
I've merged #37 into master so closing this issue for now.
Will also wait a while with merging #33 as I'd like to keep this script as basic as possible and without adding new pip dependencies if changing the headers is enough to solve the issue, please reopen a pull-request if not
